File: bots/rs/v55/src/builder/update/prune.rs
use std::collections::HashSet;

use cambc::{Controller, ControllerApi};

use crate::builder::Builder;

pub fn prune_stale(builder: &mut Builder, ct: &mut Controller<'_>) {
    builder.nearby_buildings = pyrust::vec::new!();

    builder.healable_buildings = pyrust::collect!(pyrust::filter!(pyrust::copied!(pyrust::iter!(builder
        .healable_buildings)), |p| !pyrust::unwrap!(ct.is_in_vision(*p))));
    builder.adjacent_to_enemy_launcher = pyrust::collect!(pyrust::filter!(pyrust::copied!(pyrust::iter!(builder
        .adjacent_to_enemy_launcher)), |p| !pyrust::unwrap!(ct.is_in_vision(*p))));
    builder.enemy_turret_ray_tiles = pyrust::collect!(pyrust::filter!(pyrust::copied!(pyrust::iter!(builder
        .enemy_turret_ray_tiles)), |p| !pyrust::unwrap!(ct.is_in_vision(*p))));
    builder.friendly_turret_ray_tiles = pyrust::collect!(pyrust::filter!(pyrust::copied!(pyrust::iter!(builder
        .friendly_turret_ray_tiles)), |p| !pyrust::unwrap!(ct.is_in_vision(*p))));
}

File: bots/beacon/bestrn/builder/update/prune.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from cambc import Controller, ControllerApi
if TYPE_CHECKING:
    from builder import Builder


def prune_stale(builder, ct):
    builder.nearby_buildings = []
    builder.healable_buildings = list(
        (p for p in builder.healable_buildings if not ct.is_in_vision(p))
    )
    builder.adjacent_to_enemy_launcher = set(
        (p for p in builder.adjacent_to_enemy_launcher if not ct.is_in_vision(p))
    )
    builder.enemy_turret_ray_tiles = set(
        (p for p in builder.enemy_turret_ray_tiles if not ct.is_in_vision(p))
    )
    builder.friendly_turret_ray_tiles = set(
        (p for p in builder.friendly_turret_ray_tiles if not ct.is_in_vision(p))
    )
